File: update_docs.py
#!/usr/bin/env python3

import os
import shutil
import subprocess
import sys
from pathlib import Path


def run_docs_for_llms(developers_repo_path):
    """
    Run 'pnpm docs-for-llms' in the developers.portone.io repository

    Args:
        developers_repo_path: Path to the developers.portone.io repository

    Returns:
        Path to the generated docs-for-llms directory
    """
    print(f"Running 'pnpm docs-for-llms' in {developers_repo_path}...")

    # Ensure the path exists and is a directory
    dev_repo = Path(developers_repo_path)
    if not dev_repo.exists() or not dev_repo.is_dir():
        raise ValueError(f"The provided path '{developers_repo_path}' does not exist or is not a directory")

    # Check if package.json exists to validate it's likely the correct repository
    if not (dev_repo / "package.json").exists():
        raise ValueError(f"The provided path '{developers_repo_path}' does not appear to be a valid repository (no package.json found)")

    # Run the pnpm command
    try:
        subprocess.run(["pnpm", "docs-for-llms"], cwd=str(dev_repo), check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
    except subprocess.CalledProcessError as e:
        print(f"Error running 'pnpm docs-for-llms': {e}")
        print(f"stdout: {e.stdout}")
        print(f"stderr: {e.stderr}")
        raise

    # Return the path to the generated docs directory
    generated_docs_path = dev_repo / "docs-for-llms"
    if not generated_docs_path.exists():
        raise FileNotFoundError(f"Expected generated docs at {generated_docs_path}, but directory was not found")

    return generated_docs_path


def update_mcp_docs(developers_repo_path):
    """
    Update the MCP server documentation with the latest from developers.portone.io

    Args:
        developers_repo_path: Path to the developers.portone.io repository
    """
    # Get the current script's directory
    script_dir = Path(os.path.dirname(os.path.abspath(__file__)))

    # Define the target docs directory
    target_docs_dir = script_dir / "src" / "portone_mcp_server" / "resources" / "docs"

    # Run the docs-for-llms command and get the path to the generated docs
    generated_docs_path = run_docs_for_llms(developers_repo_path)

    # Remove the existing docs directory
    print(f"Removing existing docs directory at {target_docs_dir}...")
    if target_docs_dir.exists():
        shutil.rmtree(target_docs_dir)

    # Create the parent directory if it doesn't exist
    target_docs_dir.parent.mkdir(parents=True, exist_ok=True)

    # Copy the generated docs to the target directory
    print(f"Copying new docs from {generated_docs_path} to {target_docs_dir}...")
    shutil.copytree(generated_docs_path, target_docs_dir)

    print("Documentation update completed successfully!")


def main():
    # Check if the path is provided as an environment variable
    developers_repo_path = os.environ.get("DEVELOPERS_PORTONE_IO_PATH", "")

    # If not found in environment variables, prompt the user
    if not developers_repo_path:
        print("Enter the path to the local developers.portone.io repository:")
        developers_repo_path = input().strip()

        if not developers_repo_path:
            print("Error: Repository path cannot be empty", file=sys.stderr)
            sys.exit(1)

    # Expand the tilde (~) to the user's home directory if present
    if developers_repo_path.startswith("~"):
        developers_repo_path = os.path.expanduser(developers_repo_path)

    update_mcp_docs(developers_repo_path)


if __name__ == "__main__":
    main()

File: update-docs.ts
import { exec as execCallback } from "node:child_process";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import { createInterface } from "node:readline/promises";
import { fileURLToPath } from "node:url";
import { promisify } from "node:util";
import * as yaml from "js-yaml";

const exec = promisify(execCallback);

async function runDocsForLlms(
  repoPath: string,
  nodeVersion: number,
): Promise<string> {
  console.log(`Running 'pnpm docs-for-llms' in ${repoPath}...`);

  try {
    const stats = await fs.stat(repoPath);
    if (!stats.isDirectory()) {
      throw new Error(`The provided path '${repoPath}' is not a directory`);
    }
  } catch (error) {
    if (error instanceof Error && "code" in error && error.code === "ENOENT") {
      throw new Error(
        `The provided path '${repoPath}' does not exist or is not a directory`,
      );
    }
    throw error;
  }

  try {
    const pkgPath = path.join(repoPath, "package.json");
    await fs.access(pkgPath);
  } catch (_error) {
    throw new Error(
      `The provided path '${repoPath}' does not appear to be a valid repository (no package.json found)`,
    );
  }

  try {
    const shellCommand = `. "$NVM_DIR/nvm.sh" && nvm use ${nodeVersion} && corepack enable && pnpm install && pnpm docs-for-llms`;

    const { stdout, stderr } = await exec(shellCommand, {
      cwd: repoPath,
      shell: process.env.SHELL || "/bin/bash",
      env: process.env,
    });
    if (stdout) console.log(stdout);
    if (stderr) console.error(stderr);
  } catch (e) {
    const error = e as Error & { stdout?: string; stderr?: string };
    console.error(`Error: ${error.message}`);
    if (error.stdout) console.error(`stdout: ${error.stdout}`);
    if (error.stderr) console.error(`stderr: ${error.stderr}`);
    throw error;
  }

  const generatedDocsPath = path.join(repoPath, "docs-for-llms");
  try {
    const stats = await fs.stat(generatedDocsPath);
    if (!stats.isDirectory()) {
      throw new Error();
    }
  } catch (_error) {
    throw new Error(
      `Expected generated docs at ${generatedDocsPath}, but directory was not found`,
    );
  }

  return generatedDocsPath;
}

async function fetchAndSave(
  url: string,
  targetPath: string,
  frontmatter: Record<string, any> | null,
): Promise<void> {
  console.log(`Fetching content from ${url}...`);
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), 30000);

  try {
    const response = await fetch(url, { signal: controller.signal });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }

    await fs.mkdir(path.dirname(targetPath), { recursive: true });

    const contentBuffer = Buffer.from(await response.arrayBuffer());

    if (frontmatter) {
      const frontmatterYaml = yaml.dump(frontmatter, {
        sortKeys: false,
      });
      const frontmatterBlock = Buffer.from(
        `---\n${frontmatterYaml}---\n\n`,
        "utf-8",
      );
      await fs.writeFile(
        targetPath,
        Buffer.concat([frontmatterBlock, contentBuffer]),
      );
    } else {
      await fs.writeFile(targetPath, contentBuffer);
    }

    console.log(`Content saved to ${targetPath}`);
  } catch (error) {
    console.error(`Error fetching content from ${url}: ${error}`);
  } finally {
    clearTimeout(timeoutId);
  }
}

async function updateServerSdkDocs(targetDocsDir: string): Promise<void> {
  await fetchAndSave(
    "https://raw.githubusercontent.com/portone-io/server-sdk/refs/heads/main/javascript/README.md",
    path.join(targetDocsDir, "sdk", "ko", "v2-server-sdk", "javascript.md"),
    {
      title: "포트원 V2 JavaScript, TypeScript 서버 SDK",
      description:
        "JavaScript, TypeScript를 위한 포트원 V2 서버 SDK 사용 방법을 안내합니다.",
      targetVersions: ["v2"],
    },
  );

  await fetchAndSave(
    "https://raw.githubusercontent.com/portone-io/server-sdk/refs/heads/main/jvm/README.md",
    path.join(targetDocsDir, "sdk", "ko", "v2-server-sdk", "jvm.md"),
    {
      title: "포트원 V2 JVM 서버 SDK",
      description:
        "JVM 환경(Java, Kotlin, Scala 등)을 위한 포트원 V2 서버 SDK 사용 방법을 안내합니다.",
      targetVersions: ["v2"],
    },
  );

  await fetchAndSave(
    "https://raw.githubusercontent.com/portone-io/server-sdk/refs/heads/main/python/README.md",
    path.join(targetDocsDir, "sdk", "ko", "v2-server-sdk", "python.md"),
    {
      title: "포트원 V2 Python 서버 SDK",
      description: "Python을 위한 포트원 V2 서버 SDK 사용 방법을 안내합니다.",
      targetVersions: ["v2"],
    },
  );
}

async function updateMobileSdkDocs(targetDocsDir: string): Promise<void> {
  await fetchAndSave(
    "https://raw.githubusercontent.com/portone-io/android-sdk/refs/heads/main/README.md",
    path.join(targetDocsDir, "sdk", "ko", "v2-mobile-sdk", "android.md"),
    {
      title: "포트원 V2 Android 모바일 SDK",
      description:
        "Android를 위한 포트원 V2 모바일 SDK 사용 방법을 안내합니다.",
      targetVersions: ["v2"],
    },
  );

  await fetchAndSave(
    "https://raw.githubusercontent.com/portone-io/react-native-sdk/refs/heads/main/README.md",
    path.join(targetDocsDir, "sdk", "ko", "v2-mobile-sdk", "react-native.md"),
    {
      title: "포트원 V2 React Native 모바일 SDK",
      description:
        "React Native를 위한 포트원 V2 모바일 SDK 사용 방법을 안내합니다.",
      targetVersions: ["v2"],
    },
  );
}

async function updateMcpDocs(
  developersRepoPath: string,
  helpRepoPath: string,
): Promise<void> {
  const scriptDir = path.dirname(fileURLToPath(import.meta.url));
  const targetDocsDir = path.join(scriptDir, "assets", "docs");

  console.log(`Removing existing docs directory at ${targetDocsDir}...`);
  await fs.rm(targetDocsDir, { recursive: true, force: true });

  await fs.mkdir(path.dirname(targetDocsDir), { recursive: true });

  const generatedDevDocsPath = await runDocsForLlms(developersRepoPath, 23);
  console.log(
    `Copying new docs from ${generatedDevDocsPath} to ${targetDocsDir}...`,
  );
  await fs.cp(generatedDevDocsPath, targetDocsDir, { recursive: true });

  const generatedHelpDocsPath = await runDocsForLlms(helpRepoPath, 20);
  const helpTargetDir = path.join(targetDocsDir, "help");
  console.log(
    `Copying new docs from ${generatedHelpDocsPath} to ${helpTargetDir}...`,
  );
  await fs.cp(generatedHelpDocsPath, helpTargetDir, { recursive: true });

  await updateServerSdkDocs(targetDocsDir);
  await updateMobileSdkDocs(targetDocsDir);

  console.log("Documentation update completed successfully!");
}

async function getRepoPath(
  envVarName: string,
  promptMessage: string,
): Promise<string> {
  let repoPath = process.env[envVarName] || "";

  if (!repoPath) {
    const rl = createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    console.log(promptMessage);
    repoPath = (await rl.question("")).trim();
    rl.close();
  }

  if (repoPath.startsWith("~")) {
    repoPath = path.join(os.homedir(), repoPath.slice(1));
  }

  return repoPath;
}

async function main() {
  try {
    const developersRepoPath = await getRepoPath(
      "DEVELOPERS_PORTONE_IO_PATH",
      "Enter the path to the local developers.portone.io repository:",
    );
    const helpRepoPath = await getRepoPath(
      "HELP_PORTONE_IO_PATH",
      "Enter the path to the local help.portone.io repository:",
    );
    await updateMcpDocs(developersRepoPath, helpRepoPath);
  } catch (error) {
    console.error("An unexpected error occurred:", error);
    process.exit(1);
  }
}

await main();
